text/0062-controlling-pii-and-credentials-in-sd-ks.md

```diff
@@ -0,0 +1,207 @@
+| ---        | ---                     | ---                                                                                                                                                                                                 | ---                                                                          |
+| python     | WSGI                    | [cookies](https://github.com/getsentry/sentry-python/blob/de84c4c7d65830e91f3af8555ebb1f7385a2d7cc/sentry_sdk/integrations/_wsgi_common.py#L63-L64)                                                 | all cookies including session are included which is a security issue         |
+| python     | WSGI                    | [headers](https://github.com/getsentry/sentry-python/blob/de84c4c7d65830e91f3af8555ebb1f7385a2d7cc/sentry_sdk/integrations/_wsgi_common.py#L18-L27)                                                 | hardcoded sensitive headers list                                             |
+| python     | sql spans & breadcrumbs | [params](https://github.com/getsentry/sentry-python/blob/de84c4c7d65830e91f3af8555ebb1f7385a2d7cc/sentry_sdk/tracing_utils.py#L169-L179)                                                            | not using `send_default_pii` but custom config                               |
+| javascript | httpclient              | [cookies, headers](https://github.com/getsentry/sentry-javascript/blob/7316b850988df64467a7052082db84461f8a2cd2/packages/integrations/src/httpclient.ts#L136-L154)                                  | no control in which cookies and headers                                      |
+| ruby       | request interface       | [headers](https://github.com/getsentry/sentry-ruby/blob/bbfb6514d95adda77676d4a0823924e789db0909/sentry-ruby/lib/sentry/interfaces/request.rb#L84-L104)                                             | random hardcoded lists                                                       |
+| android    | okhttp                  | [headers](https://github.com/getsentry/sentry-java/blob/9b6c37f33376556ce0f66e526541a04e9e3aa60f/sentry-android-okhttp/src/main/java/io/sentry/android/okhttp/SentryOkHttpInterceptor.kt#L223-L242) | no headers at all if `send_default_pii` false, filter sensitive even if true |
+
+As demonstrated by the collection above, each SDK does slightly different things and the decisions on what is gated behind `send_default_pii` are mostly left to the integration authors.  
+The user can merely choose to follow the author's implementation with this flag but not really be sure if and what data is leaking without reading the integration source, which is an undesirable state of affairs.
+
+
+#### `before_send` and `add_event_processor`
+
+In cases where users are not happy with `send_default_pii`, we suggest using either `before_send` for errors or `add_event_processor` for all events for controlling PII on the final event payload.  
```

Comment:
We seem to be fighting the symptom of not knowing which PIIs our SDKs are sending instead of solving the underlying problem. Users should trust our SDKs not to transmit any PII data to Sentry per default. The SDK maintainers are fully responsible for the data our SDKs send to Sentry. Asking our users to implement some of the options below is loading the problem off to them. How can they be sure a newly added feature by us doesn't start sending PII? Do they have to adapt their logic for new features? Even with the endeavors below, we can't guarantee to scrub all PII from our payloads.
Users are responsible if they manually add any PII to Sentry data. We can't check all their data to ensure they don't send PII to us. Of course, we could go the extra mile to implement one of the options, but that's just being polite, and I wonder if that's required.
Instead, we have to go through every SDK to ensure they don't send any PII if send_default_pii is off. Furthermore, we have to find a way to ensure no new feature starts transmitting PII by accident.
One step in this direction could be to have a checkbox in PR templates to validate that no PII gets sent by accident. We recently added this to some of the mobile SDKs: https://github.com/getsentry/sentry-cocoa/pull/2651/files

Reply:
we don't always know what is PII in a user's application?

Reply:
@sl0thentr0py, but we know which data we collect automatically and send to Sentry. If we can't ensure it doesn't contain PII, we should only send it if `send_default_pii` is on. There is a big difference between the user manually actively passing PII to Sentry, like `scope.user.email = 'my@user.com'` , or we intercepting HTTP requests and collecting headers and parameters.

Reply:
@philipphofmann I believe it's part of the exercise to double-check what we send and what we can catch.
Adding a checkbox to the PR template at least reminds you to think about it.
As always, a 100% guarantee doesn't exist, we can try our best tho.

Reply:
> If we can't ensure it doesn't contain PII, we should only send it if send_default_pii is on.

this is impractical, headers or stack variables can always potentially contain sensitive stuff. So we'd have to just stop sending headers in this case, which is a severe functionality regression.

> Asking our users to implement some of the options below is loading the problem off to them. 

We are not asking them to implement anything, everything in this RFC has defaults that then can be _easily_ tweaked in increasing complexity if they so wish to do so.

I'm not sure if the proposal came across effectively, so I will attempt a simpler version.

The current control for PII is `send_default_pii`. This RFC expands those controls to a combination of `send_default_pii` and a new `denylist` (that has defaults). 

This way the user can *additionally* let us know what in *their application* could be PII on top of what we think is PII. This is a more comprehensive handling of the state of affairs than currently exists. We are expanding the semantics of what is PII and giving the user a say in choosing those semantics.

Everything else in this RFC is implementation details for expanding that basic semantic.

Reply:
> this is impractical, headers or stack variables can always potentially contain sensitive stuff. So we'd have to just stop sending all headers in this case, which is a severe functionality regression.

Can we narrow it down to headers that don't contain any PII? If not, we have to ensure to scrub all data out of the headers to not send any PII. 

> We are not asking them to implement anything, everything in this RFC has defaults that then can be easily tweaked in increasing complexity if they so wish to do so.

Sorry, my bad, then. It was late, and maybe I didn't read the RFC thoroughly.

> This way the user can additionally let us know what in their application could be PII on top of what we think is PII. This is a more comprehensive handling of the state of affairs than currently exists. We are expanding the semantics of what is PII and giving the user a say in choosing those semantics.

What I don't like about this approach is that even if `send_default_pii` is disabled, we will send PII to Sentry if users put PII into HTTP headers, not following a proper naming pattern and even if we have good data scrubbers. For example, how can we ensure users are not doing
```
Nammee-Addressi: Philipp Hofmann And My Address and because I don't know what I'm during here is also my social security number 123456789
```

Why don't we instead only collect headers not meant to contain PII? So basically go with an allowlist of HTTP headers that users can add headers to? Like for example:

```
Age: 2318192
Cache-Control: public, max-age=315360000
Connection: keep-alive
Date: Mon, 18 Jul 2016 16:06:00 GMT
Server: Apache
Vary: Accept-Encoding
X-Backend-Server: developer6.webapp.scl3.mozilla.com
X-Cache: Hit from cloudfront
...
```

Reply:
> Can we narrow it down to headers that don't contain any PII? 

No we cannot which is why we let users tell us what is PII via a `denylist`.

> how can we ensure users are not doing

we cannot, which is why we rely on users to tell us and add `Namee-Addressi` to the `denylist`.

>   basically go with an allowlist

the default of product design should be adding value (which here is having most headers by default) with the _optionality_ to remove when it is not appropriate. The `allowlist` is taking away the feature by default and selectively turning it on.

I will stick to the `denylist`, others are welcome to chime in on what they prefer.

Reply:
> the default of product design (and a big reason our SDKs are successful) should be adding value

Yes, but privacy and security should be more important than customer value. Is it clear to customers that we send PII if they disable `send_default_pii`? How common is it that PII ends up in HTTP headers?

Your suggested approach might work if sending PII is an edge case. So the approach would be that we try to avoid sending PII, but we can't guarantee that. Here is the list of features that might collect PII, so please double-check. You can configure an extra deny list to ensure PII doesn't end up in Sentry. If the docs explain this, it might be acceptable.

Reply:
> Yes, but privacy and security should be more important than customer value.
Why would someone use any service that doesn't add value? I feel like this is not a good argument to make here.

Having a pragmatic way of removing Pii/sensitive data from the data we collect, or set by the user, seems to be missing in today's SDKs.
A on/off toggle doesn't cut it, hence the approach outlined in this RFC is necessary.

Our users should still be able to send us their request headers/bodies, with an easy-to-use API that can perform client-side scrubbing.

Reply:
>  If it's turned off, and we still send PII

The name of the option is `send_default_pii`, not `send_pii`. Without the user's input about what PII is in *their application*, we can not *deterministically* tell ourselves what is PII or not.

> Off. We don't send any PII, period.

This is not possible with the current state of affairs, even if you remove headers and such. We are trying to improve the current handling, but we cannot still end up in a system where only a boolean controls this. The only way to guarantee this 100% is to not use Sentry.

Reply:
> The name of the option is send_default_pii, not send_pii. Without the user's input about what PII is in their application, we can not deterministically tell ourselves what is PII or not.

I'm confused by the name, then. What's default PII?

> Off. We don't send any PII, period.

I'm only referring to data we automatically collect. If the user sends an email address to the scope, it's their responsibility. 

Maybe it's better to hop on a call, @sl0thentr0py, to discuss this.

Reply:
I will add it to TSC.